### ISSUE TITLE
[Feature Request] Headless mode (to run on server)

### ISSUE THREAD

**Issue by gboudrias** [crowbartools/Firebot#2227]
**Describe the solution you'd like**
I would like to run this on a cloud server in order for events to be registered even if my computer is not on. Since the config already lives in a file, I'm assuming the main part that's missing is a way to launch without the UI.

**Additional context**
I might try and submit a PR if I can figure something out, any clues in that regard would be appreciated.

**Comment by lublak** [comment]
It would also be interesting if you could connect to the server anyway.
Either via the browser http://firebotserver/ or as remote access via the application:
Control headless instance, give server address:
firebot:939

**Comment by kawazoe** [comment]
Chipping in, I'm a mod on a friend's stream and we would really like to stop using other cloud-based solution in favor of something self-hosted. If I could run Firebot as a docker container, as suggested in #2780, I could easily setup a web server for it that both me and the streamer could access during a stream. Firebot already does everything we need besides letting mods configure the bot.

**Comment by smarithor** [comment]
Chipping in, similar to Kawazoe. I mod for a couple of streamers that use Firebot and anytime I need to help them set anything up I need to remote into their machines to get access to Firebot. Having Firebot running as a headless server with internet access would be nice so I can just connect to that instead of having to connect to their machines.

**Comment by Kuroo2021** [comment]
Yeah, I’d really love to see this feature added. A headless or server mode would make Firebot way more flexible for remote streaming setups. It would also help a lot when mods or collaborators need access without remote desktop tools. I can see it being really useful for managing events and commands from anywhere. I’d definitely use this if it becomes available.

**Comment by zunderscore** [comment]
Just chiming in here:

This is something we absolutely plan on doing, but it will require a completely different architecture than we have today, so we won't be able to do this until the full rewrite in v6, which we have no ETA for.

**Comment by kawazoe** [comment]
I've spent some energy on the problem and found half of a hack, and a disappointing solution.

## The half hack

Since I'm not looking into customizing the UI or use Firebot as an API, I thought of a nice piece of kit from https://github.com/jlesage/docker-baseimage-gui which lets you host a full GUI app in a docker container running a lightweight window manager that renders directly to a VNC stream. You can then either use a VNC client to connect to the app, or the built-in VNC web server to reach the app via a browser. Since it's just VNC, you can also have multiple people interacting with the GUI at the same time! (kind-of)

> Yes, I am aware of the insanity of running what is functionally node, react, and chrome just to render to a fake DE that streams the output over VNC to capture it and render it to a canvas from a web server all hosted literally on the same machine.

This would have covered my needs. I already do have other apps that use this base image and is actually how I discovered it existed in the first place.

I gave up when I noticed installing the deb release of Firebot came with 436 transient dependencies, including a full KDE install with sounds and icons, SystemD, and even iOS specific debug tools, and most of that stuff simply refuses to install in a docker container. If someone wants to continue pushing in that direction, you can follow the instruction at the link up there and see how far you get. Getting SystemD working in docker is way beyond my comfort zone.

## The disappointing solution

I spun up a "lightweight" debian compatible distro in a VM on my NAS and installed Firebot directly there. It's easy enough to connect to it so there's that. We can't both use it at the same time, but at least we can timeshare it which is an improvement from running it on a machine I cannot access at all. It uses around 8 GB of RAM which is probably worth a house and a half but hey, it works enough to wait for v6.

**Comment by KayOhtie** [comment]
> I gave up when I noticed installing the deb release of Firebot came with 436 transient dependencies, including a full KDE install with sounds and icons, SystemD, and even iOS specific debug tools, and most of that stuff simply refuses to install in a docker container. If someone wants to continue pushing in that direction, you can follow the instruction at the link up there and see how far you get. Getting SystemD working in docker is way beyond my comfort zone.
> ## The disappointing solution
> 
> I spun up a "lightweight" debian compatible distro in a VM on my NAS and installed Firebot directly there. It's easy enough to connect to it so there's that. We can't both use it at the same time, but at least we can timeshare it which is an improvement from running it on a machine I cannot access at all. It uses around 8 GB of RAM which is probably worth a house and a half but hey, it works enough to wait for v6.

Wait why do all of that when [xpra](https://xpra.org/index.html) exists and can just... share from whatever you normally run it on to other machines, including multiple simultaneously? That's what I ended up doing, at least when I'm not just using Rustdesk to remote to my PC.

**Comment by kawazoe** [comment]
> Wait why do all of that when [xpra](https://xpra.org/index.html) exists and can just... share from whatever you normally run it on to other machines, including multiple simultaneously? That's what I ended up doing, at least when I'm not just using Rustdesk to remote to my PC.

Because the remoting protocol doesn't change my requirements.

I don't want to remote into the streamer's computer. Sometimes I need to work on the bot while they're not streaming or even have their computer on. Can you imagine if your discord bot only works when you're on your computer? In any serious community, that stuff needs to be up 24/7.

We also both have a full wayland setup and I do not know how xpra would behave there and am not interested to figure it out either. I want to host the app in a shared, always-on environment and that one can have an X11 window manager if it needs to. That shared environment can use either TigerVNC or xpra. It really doesn't matter.

However, I also want to be able to access the app from my phone or tablet, which only leaves me with these options: an Android or iOS VNC client, an HTML5 VNC client hosted on the shared environment, or an xpra HTML5 client hosted on the shared environment.

In the end, they all have the same problem. If the shared environment is a docker container, you can't install the deb because of the systemd dependency. It needs to be a VM, and at that point, I can just remote into it with a browser already.
